Fix fuzzer instances stopping during parallel fuzzing

arata-nvm opened the pull request:
<!-- Set a title that summarizes the PR changes. -->

## Type of PR
<!-- This project uses issue-driven development, so please take the appropriate action for each PR type. -->

- [ ] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [x] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Related Issue
<!-- If this PR is a PR type A/B-1, please provide a link to the corresponding issue. -->
<!-- If this PR is a PR type B-1, please write "N/A" -->

## Importance of PR
<!-- Please describe the importance of the PR in terms of the following aspects. -->

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [ ] Medium (requires several hours to a day of review)
    - [x] Small (trivial changes, such as build error)

## PR Overview
<!-- Please provide a summary of this PR. -->
<!-- If this PR is a PR type A/B-1, this PR will be considered as an item in the checklist for the related issue. Please provide a link to the issue comment that contains the checklist. -->
<!-- If this PR is a PR type B-2, unnecessary to reference the issue. Please provide a summary. -->

Currently, the AFL family of fuzzuf have a bug that fuzzer instances randomly stop after a while.

I found two causes for this bug:

1. `mmap` returns an error when trying to read an empty file. This is fixed by skipping the file when attempting to read an empty one.

2. A mmapped file can be modified by another process. This happens, for example, when importing test cases from other instances during parallel fuzzing. This is fixed by calling `unlink` before overwriting the test case file. Since AFL++ has an cli option to toggle whether to call `unlink` or not, I implemented it in the same way.

Several fuzzers (afl, aflfast, aflplusplus, afl_symcc, die, ijon, mopt, rezzuf, eclipser) are affected by this fix.

I confirmed that instances do not stop for 24 hours when fuzzing [exifutil](https://github.com/fuzzuf/fuzzuf/tree/master/docs/resources/exifutil) using the following commands:

```shell
screen -dmS exifutil-master fuzzuf afl -i /corpus -o /eval -M master -- exifutil -f @@
for i in `seq 1 32`; do
  screen -dmS exifutil-fuzzer${i} fuzzuf afl -i /corpus -o /eval -S fuzzer${i} -- exifutil -f @@
done
```

## Concerns (Optional)
<!-- If you have any concerns, please describe them clearly by filling in the relevant checklist items below. If there is anything else you would like to share with the reviewer, please include it. -->

- [ ] Performance
- [ ] Source Code Quality

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [ ] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [x] The PR title is a summary of the changes.
- [x] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

arata-nvm committed:
feat: add `--no-unlink` cli option for AFL family

arata-nvm committed:
feat: unlink files before overwriting

arata-nvm committed:
feat: add parameter to ignore ENOENT when mapping files

arata-nvm committed:
feat: skip empty testcase

retrage commented on utils/map_file.cpp:
```diff
@@ -56,6 +60,10 @@ auto map_file(const std::string &filename, unsigned int flags, bool populate)
     file_size = stat_.st_size;
   }
 
+  if (file_size == 0) {
```
nit: Add a comment
```suggestion
  // Skip if empty to avoid mmap error
  if (file_size == 0) {
```

retrage commented on include/fuzzuf/utils/mmap_range.hpp:
```diff
@@ -78,9 +78,9 @@ class mmap_iterator /*<T,
                             typename std::iterator_traits<T>::value_type>,
                         fs::path>) {
-        cur = map_file((*base).string(), flags, populate);
+        cur = map_file((*base).string(), flags, populate, true);
```
Add a comment on enabling `ignore_enoent` as it's a special case. 
```suggestion
        // NOTE: ignore ENOENT error because the file may be unlinked.
        cur = map_file((*base).string(), flags, populate, true);
```

retrage reviewed (commented):
The overall proposed changes look good to me. Some trivial comments:

## Add `-N` option description to the docs

Add a description to `docs/algorithms/afl/algorithm_en.md` something like:
```
- Local options (only available for AFL)
    - `--dict_file=path/to/dict/file`
        - Specifies a path to the file, loaded as an additional dictionary.
    - `--no-unlink`
        - Do not unlink the fuzzing input file.
```

## Other algorithms that support parallel fuzzing

`afl_symcc` and `eclipser` support AFL-style parallel fuzzing, but the `-N` option is not added to them. Is it intentional?

arata-nvm committed:
docs: add -N option description to the docs

Co-authored-by: Akira Moroo <akiram@ricsec.co.jp>

arata-nvm commented:
Thank you for your review.

## Add -N option description to the docs

Fixed in https://github.com/fuzzuf/fuzzuf/pull/122/commits/c3ff724d32aaad5c57c4f4a7aa43dd0acb69cb2d .

## Other algorithms that support parallel fuzzing

It is not intentional. I added the -N option to afl_symcc(https://github.com/fuzzuf/fuzzuf/pull/122/commits/109ae5c440b5d8eea07ae9c0cced4380cd815598) and eclipser(https://github.com/fuzzuf/fuzzuf/pull/122/commits/39d8a3af23ae0f10e0b0d0b590f32c0dbd223321, https://github.com/fuzzuf/fuzzuf/pull/122/commits/1f4595a13adc78226665eb8950508d2fca58454d).

retrage reviewed (approved):
It looks good to me. Please do `git rebase` to clean up the commit history.
